### PR TITLE
(Updated) #1083: use unstructured runtime.Object for more detailed reflector logging

### LINKED ISSUE #1083: Informer logging lacks type detail
KubeFed's sync controller uses informers to watch federated resources in the host cluster and target resources in member clusters. The sync controller is configured with type information via the API to support the federation of types (e.g. CRDs) for which golang structs may not be available at compilation time.  Accordingly, the sync controller configures its informers with an `ObjectType` (indicating the type of resource that will be watched) of `nil`. `ObjectType` is passed by an informer to an internal `Reflector` instance, where it is used to:

1 -  check that the type of an event's object is as expected (i.e. [1](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L345))
2 -  indicate the type of resource that a log pertains to (e.g. [1](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L122), [2](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L201))

Passing `nil` for `ObjectType` avoids the type check to support watching resources for which a golang struct is not available, but also results in most `Reflector` log messages displaying `<nil>` instead of a type name: 

```
0806 13:49:50.180103    3856 reflector.go:131] Starting reflector <nil> (0s) from sigs.k8s.io/kubefed/pkg/controller/sync/accessor.go:161
I0806 13:49:50.180207    3856 reflector.go:169] Listing and watching <nil> from sigs.k8s.io/kubefed/pkg/controller/sync/accessor.go:161
I0806 13:49:50.180106    3856 reflector.go:169] Listing and watching <nil> from sigs.k8s.io/kubefed/pkg/controller/sync/accessor.go:156
```

Ideally `Reflector` could be configured with a `typeName` to be used in log messages. If `typeName` were not provided, it would be continue to be computed as `expectedType.String()`. A new constructor for `Reflector` and `Informer` could be added to support provision of `typeName`.

Note that KubeFed would likely prefer to supply a `typeName` of `<cluster name>/<type name>` to ensure traceability of an informer not just to a sync controller but also to the cluster it is targeting.

### PR DESCRIPTION
**Pull request by sreis**
**What this PR does / why we need it**:

Rebases #1146 with master to bring it up to date. We've used that PR successfully to track down a couple of problems and didn't found any issues while using it.

Close #1146 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #1083

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/controller/util/resourceinformer.go [hunk 1]
```diff
@@ -52,7 +60,7 @@ func newResourceInformer(client ResourceClient, namespace string, triggerFunc fu
 				return client.Resources(namespace).Watch(options)
 			},
 		},
-		nil, // Skip checks for expected type since the type will depend on the client
+		obj, // use a unstructured type with apiVersion / Kind populated for informer logging purposes
```
Comment (jimmidyson) [inline pkg/controller/util/resourceinformer.go]: nit: 
```suggestion
		obj, // use an unstructured type with apiVersion / kind populated for informer logging purposes
```

File: pkg/controller/util/resourceinformer.go [hunk 2]
```diff
@@ -23,24 +23,32 @@ import (
...
 }
 
-func newResourceInformer(client ResourceClient, namespace string, triggerFunc func(pkgruntime.Object), labelSelector string) (cache.Store, cache.Controller) {
+func newResourceInformer(client ResourceClient, namespace string, apiResource *metav1.APIResource, triggerFunc func(pkgruntime.Object), labelSelector string) (cache.Store, cache.Controller) {
+	obj := &unstructured.Unstructured{}
+
+	if apiResource != nil {
+		gvk := schema.GroupVersionKind{Group: apiResource.Group, Version: apiResource.Version, Kind: apiResource.Kind}
+		obj.SetGroupVersionKind(gvk)
+		obj.SetNamespace(namespace)
```
Comment (jimmidyson) [inline pkg/controller/util/resourceinformer.go]: Is this used by anything?

Comment (sreis) [reply pkg/controller/util/resourceinformer.go]: Good question. I don't think it does but I'll do some manual test to confirm and report back.

Comment (xunpan) [reply pkg/controller/util/resourceinformer.go]: Could you explain more about this `SetNamespace()`?  If it does not needed for current logic, keep it here makes less sense.

Comment (jimmidyson) [reply pkg/controller/util/resourceinformer.go]: I've read through the code and this is unused afaict. It doesn't affect the logging so removing it is fine.

**Pull request merged**